sql/sql_admission_control.h

```diff
@@ -0,0 +1,356 @@
+/* Copyright (c) 2016, Facebook. All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#ifndef _sql_admission_control_h
+#define _sql_admission_control_h
+
+#include "mysql/psi/mysql_mutex.h"
+#include "mysql/psi/psi_base.h"
+
+#include <deque>
+#include "sql/sql_class.h"
+
+#define IS_BIT_SET(val, n) ((val) & (1 << (n)))
+
+extern bool opt_admission_control_by_trx;
+extern ulonglong admission_control_filter;
+class AC;
+extern AC *db_ac;
+
+#ifdef HAVE_PSI_INTERFACE
+extern PSI_stage_info stage_admission_control_enter;
+extern PSI_stage_info stage_admission_control_exit;
+extern PSI_stage_info stage_waiting_for_admission;
+#endif
+
+// TODO: remove this block eventually when this becomes part of audit plugin?
+// TBD later
+// Resource isolation types that a multi-tenancy plugin will handle.
+// Currently connection and query limits are two resource types. More will be
+// supported in the future.
+enum class enum_multi_tenancy_resource_type : int32_t {
+  MULTI_TENANCY_RESOURCE_CONNECTION,
```

Comment:
enum classes always require type name, thus the value name prefixes can be dropped: CONNECTION, QUERY, NUM_RESOURCE_TYPES

Reply:
ok, just read about "enum classes". Interesting addition in C++11. But..

This enum will be removed when the code moves to separate plugin. It is just there to keep the changes to minimal(when compared with 5.6 patch).

Will add TODO here